uPortal-api/uPortal-api-rest/src/test/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServletTest.java

```diff
@@ -110,7 +259,109 @@ public void testRemoveFavorite() throws IOException {
         person.setUserName("jdoe");
         person.setFullName("john doe");
         IUserInstance userInstance = new UserInstance(person, null, null);
-        Mockito.when(userInstanceManager.getUserInstance(req)).thenReturn(userInstance);
-        ModelAndView modelAndView = updatePreferencesServlet.removeFavorite("channelId", req, res);
+        when(userInstanceManager.getUserInstance(req)).thenReturn(userInstance);
+        updatePreferencesServlet.removeFavorite("channelId", req, res);
+    }
+
+    @Test
+    public void testAddFavoriteSucceeds() throws IOException {
+        req.setLocalName("en-US");
+        updatePreferencesServlet.setPersonEntityService(
+                this.createPersonEntityService(personEntity));
+        whenFavoritesFolderColumnNodeIsPresent();
+        whenFavoritedPortletsArePresent(
+                favoritedPortlet1Id, favoritedPortlet2Id, favoritedPortlet3Id);
+        ModelAndView mav =
+                updatePreferencesServlet.addFavorite(idForPortletBeingFavorited, req, res);
+        assertNotNull(mav);
+        assertEquals("jsonView", mav.getViewName());
+        assertEquals(idForPortletBeingFavorited, mav.getModel().get("newNodeId"));
+        verify(ulm).addNode(any(), eq(favoritesFolderColumnNodeId), eq((String) null));
+        verify(ulm).saveUserLayout();
+    }
+
+    @Test
+    public void testAddFavoriteRemovesFavoriteForPortletThatHasBeenDeleted() throws IOException {
+        req.setLocalName("en-US");
+        updatePreferencesServlet.setPersonEntityService(
+                this.createPersonEntityService(personEntity));
+        whenFavoritesFolderColumnNodeIsPresent();
+        whenFavoritedPortletsArePresent(
+                favoritedPortlet1Id, MissingPortletDefinition.CHANNEL_ID, favoritedPortlet3Id);
+        ModelAndView mav =
+                updatePreferencesServlet.addFavorite(idForPortletBeingFavorited, req, res);
+        assertNotNull(mav);
+        assertEquals("jsonView", mav.getViewName());
+        assertEquals(idForPortletBeingFavorited, mav.getModel().get("newNodeId"));
+        verify(ulm).deleteNode(orphanedFavoritedPortletSubscribeId);
+        verify(ulm).addNode(any(), eq(favoritesFolderColumnNodeId), eq((String) null));
+        verify(ulm).saveUserLayout();
+    }
```

Comment:
Would it also make sense to test for an error case like null id?
Something like:

```java
@Test
public void testAddFavoriteWithNullPortletId() throws IOException {
    MockHttpServletRequest request = new MockHttpServletRequest();
    MockHttpServletResponse response = new MockHttpServletResponse();
    
    ModelAndView mav = updatePreferencesServlet.addFavorite(null, request, response);
    
    assertNotNull(mav);
    assertEquals("errorView", mav.getViewName());
    verify(logger).error("Invalid input parameters for adding favorite.");
}
```

Reply:
Sure, I added a test for null id.